device/transport/amqp/src/amqp.ts

```diff
@@ -74,54 +83,331 @@ export class Amqp extends EventEmitter implements Client.Transport, StableConnec
+          },
+          attachC2DLink: (callback) => {
+            debug('attaching C2D link');
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_030: [The `Amqp` object shall attach the C2D `ReceiverLink` object if necessary to start receiving messages.]*/
+            this._amqp.attachReceiverLink(this._c2dEndpoint, null, (err, receiverLink) => {
+              if (err) {
+                debug('error creating a C2D link: ' + err.toString());
+                handleResult('AMQP Transport: Could not attach link', callback)(err);
+              } else {
+                debug('C2D link created and attached successfully');
+                this._c2dLink = receiverLink;
+                this._c2dLink.on('error', this._c2dErrorListener);
+                this._c2dLink.on('message', this._c2dMessageListener);
+                callback();
+              }
+            });
+          },
+          detachC2DLink: (callback) => {
+            this._stopC2DListener(undefined, callback);
+          },
+          onDeviceMethod: (methodName, methodCallback) => {
+            /*Codes_SRS_NODE_DEVICE_AMQP_16_022: [The `onDeviceMethod` method shall call the `onDeviceMethod` method on the `AmqpDeviceMethodClient` object with the same arguments.]*/
+            this._deviceMethodClient.onDeviceMethod(methodName, methodCallback);
+            this._deviceMethodClient.attach((err) => {
+              if (err) {
+                debug('error while attaching the device method client: ' + err.toString());
+                // no need to emit anything: the general error handler registered in the constructor will do that for us.
+              }
+            });
+          }
+        },
+        disconnecting: {
+          _onEnter: (disconnectCallback, err) => {
+            let finalError = err;
+            async.series([
+              (callback) => {
+                if (this._d2cLink) {
+                  let tmpD2CLink = this._d2cLink;
+                  this._d2cLink = undefined;
+
+                  if (err) {
+                  /*Codes_SRS_NODE_DEVICE_AMQP_16_023: [The `disconnect` method shall forcefully detach all attached links if a connection error is the causing the transport to be disconnected.]*/
+                    tmpD2CLink.forceDetach(err);
+                    tmpD2CLink.removeListener('error', this._d2cErrorListener);
+                  }
+                  tmpD2CLink.detach((detachErr) => {
+                    if (detachErr) {
+                      debug('error detaching the D2C link: ' + detachErr.toString());
+                    } else {
+                      debug('D2C link detached');
+                    }
+                    tmpD2CLink.removeListener('error', this._d2cErrorListener);
+                    if (!finalError && detachErr) {
+                      finalError = translateError('error while detaching the D2C link when disconnecting', detachErr);
+                    }
+                    callback(finalError);
```

Comment:
>callback(finalError); [](start = 20, length = 21)

Does this abort the series?  If so, what's the point of the async.series?

Reply:
correct. my bad, shouldn't call its callback with the finalError.

---
In reply to: [141106618](https://github.com/Azure/azure-iot-sdk-node/pull/89#discussion_r141106618) [](ancestors = 141106618)